checks/shell_download_validate.go

```diff
@@ -0,0 +1,628 @@
+	in := strings.NewReader(cmd)
+	f, err := syntax.NewParser().Parse(in, "")
+	if err != nil {
+		return false, ErrParsingShellCommand
+	}
+
+	cmdValidated := true
+	syntax.Walk(f, func(node syntax.Node) bool {
+		if isFetchPipeExecute(node, cmd, pathfn, logf) {
+			cmdValidated = false
+		}
+
+		// Continue walking the node graph.
+		return true
+	})
+
+	return cmdValidated, nil
+}
+
+func recordFetchFileFromNode(node syntax.Node) (pathfn string, ok bool, err error) {
+	ss, ok := node.(*syntax.Stmt)
+	if !ok {
+		return "", false, nil
+	}
+
+	cmd, ok := extractCommand(ss.Cmd)
+	if !ok {
+		return "", false, nil
+	}
+
+	if !isDownloadUtility(cmd) {
+		return "", false, nil
+	}
+
+	fn, ok := getRedirectFile(ss.Redirs)
+	if !ok {
+		return getOutputFile(cmd)
+	}
+
+	return fn, true, nil
+}
+
+func validateCommandIsNotFetchToFileExecute(cmd, pathfn string, logf func(s string, f ...interface{})) (bool, error) {
+	in := strings.NewReader(cmd)
+	f, err := syntax.NewParser().Parse(in, "")
+	if err != nil {
+		return false, ErrParsingShellCommand
+	}
+
+	cmdValidated := true
+	files := make(map[string]bool)
+	syntax.Walk(f, func(node syntax.Node) bool {
+		if err != nil {
+			return false
+		}
+		var fn string
+		var b bool
+		// Record the file that is downloaded, if any.
+		fn, b, err = recordFetchFileFromNode(node)
+		if err != nil {
+			return false
+		} else if b {
+			files[fn] = true
+		}
+
+		// Check if we're calling a file we previously downloaded.
+		if isExecuteFiles(node, cmd, pathfn, files, logf) {
+			cmdValidated = false
+		}
+
+		// Continue walking the node graph.
+		return true
+	})
+
+	return cmdValidated, err
+}
+
+func isFetchStdinExecute(node syntax.Node, cmd, pathfn string,
+	logf func(s string, f ...interface{})) bool {
+	ce, ok := node.(*syntax.CallExpr)
+	if !ok {
+		return false
+	}
+
+	c, ok := extractCommand(ce)
+	if !ok {
+		return false
+	}
+
+	if !isInterpreter(c) {
+		return false
+	}
+
+	// Now parse the process substitution part.
+	// Example: `bash <(wget -qO- http://website.com/my-script.sh)`.
+	l := 2
+	if len(ce.Args) < l {
+		return false
+	}
+
+	parts := ce.Args[1].Parts
+	if len(parts) != 1 {
+		return false
+	}
+
+	part := parts[0]
+	p, ok := part.(*syntax.ProcSubst)
+	if !ok {
+		return false
+	}
+
+	if !strings.EqualFold(p.Op.String(), "<(") {
+		return false
+	}
+
+	if len(p.Stmts) == 0 {
+		return false
+	}
+
+	c, ok = extractCommand(p.Stmts[0].Cmd)
+	if !ok {
+		return false
+	}
+
+	if !isDownloadUtility(c) {
+		return false
+	}
+
+	logf("!! frozen-deps/fetch-execute - %v is fetching and executing non-pinned program '%v'",
+		pathfn, cmd)
+	return true
+}
+
+func isSuCommand(cmd []string) bool {
+	isSu := false
+	for _, c := range cmd {
+		if strings.EqualFold("su", c) {
+			isSu = true
+		} else if isSu && strings.HasPrefix(c, "-") && strings.Contains(c, "c") {
+			return true
+		}
+	}
+	return false
+}
+
+func extractSuCommand(args []*syntax.Word) (string, bool) {
```

Comment:
some more comments on what the expected input/output is, what the functions are doing in this file would make this much more easier to understand in general.

Reply:
I'll add comments in a follow-up PR. I agree I was light on comments :/

Reply:
thanks!